### PR TITLE
Features/rental assistance part 2 #158503202

### PR DESCRIPTION
**Pull request by fddayan**

### CODE REVIEW

File: config/routes.rb
Comment (callaghanc) [inline config/routes.rb:51]: This is the only important changed line relevant to rental assistances in this file. The rest of the changes are just tidying and alphabetization.

File: app/javascript/components/supplemental_application/sections/RentalAssistance.js
Comment (callaghanc) [inline app/javascript/components/supplemental_application/sections/RentalAssistance.js]: This class and the others added below this in this file are added to enable us to test. The e2e tests need identifiers to be able to locate and interact with the rental assistance form components.

File: spec/javascript/e2e/ApplicationNewPage.e2e.js
Comment (callaghanc) [inline spec/javascript/e2e/ApplicationNewPage.e2e.js:37]: All of these `await`s that you see added in this PR before the `browser.close()`s in our e2e tests are to ensure that we actually wait for the browser to close! Not having these awaits is what was causing the browsers in the e2e tests to never actually close and causing the whole test process to hang for a long time after finishing the tests.

**Comment by callaghanc** [comment]
This PR is code-complete, but we are going to hold off on reviewing this until part 1 is fully QA'ed and merged to master, so this PR will remain WIP for now.

**Comment by callaghanc** [comment]
This PR is ready for review.

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (akegan) [inline app/javascript/components/supplemental_application/SupplementalApplicationPage.js]: does this already display a message or do you have to pass one in?

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (akegan) [inline app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: do you think that it'd be possible to condense the create/update/destroy or are there enough slight differences that it's not worth it?

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (akegan) [inline app/javascript/components/supplemental_application/SupplementalApplicationPage.js:370]: why doesn't rentalAssistances need to use mapList?

File: app/services/force/soql/rental_assistance_service.rb
Comment (akegan) [inline app/services/force/soql/rental_assistance_service.rb:1]: so many new clean and organized services! 🎉

File: spec/fixtures/model/force/lease/lease_domain.json
Comment (akegan) [inline spec/fixtures/model/force/lease/lease_domain.json]: I thought domain was snake_case? I believe other domain fixtures are.

File: spec/fixtures/model/force/lease/lease_salesforce_from_domain.json
Comment (akegan) [inline spec/fixtures/model/force/lease/lease_salesforce_from_domain.json:1]: does this file name need the "from_domain" or is it just lease_salesforce?

File: spec/fixtures/model/force/rental_assistance/rental_assistance_salesforce_from_domain.json
Comment (akegan) [inline spec/fixtures/model/force/rental_assistance/rental_assistance_salesforce_from_domain.json:1]: same here re: filename

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (akegan) [inline spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: nit: this section might be bit clearer if the page.$evals are grouped before or after the recurring/amount.

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (akegan) [inline spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: why is it nth-last-child(2) if it's the last row?

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (akegan) [inline spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: at some point soon we may want to think about breaking this up into multiple files so it doesn't get too long. For now, do you think it might help to add another layer of "describe" to help organize the e2e tests a bit?

**Review by akegan — COMMENTED** [review]
I think these are all only comments/questions. Looks great! I really appreciate all the cleanup on the backend and the e2e tests!

also hooray for adding the arrow down! 🎉

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (callaghanc) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js]: this triggers our standard window alert, which has the message "Oops! Looks like something went wrong. Please try again."

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (callaghanc) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: there are enough differences in the post-request steps that it's probably not worth squishing them into a single method; however, perhaps the loading state management and error alert could be DRYed a bit - @fddayan what are your thoughts?

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (callaghanc) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:370]: because the rental assistances already come in in domain format! 🎉 thanks to our refactor work, the SOQL rental assistance service uses SOQL to fetch the rental assistances, and then translates them to domain before passing them up to the controller. that way the controller can just hand the frontend domain-formatted rental assistances. the ultimate goal here is to do away with all use of `mapList` in the frontend. and then, even further then that, once that's done and the frontend is entirely in domain, we can consider whether we'd like to convert domain format from snake case to camel case - but that's a little further off yet.

File: app/services/force/soql/rental_assistance_service.rb
Comment (callaghanc) [reply app/services/force/soql/rental_assistance_service.rb:1]: 😍

File: spec/fixtures/model/force/lease/lease_domain.json
Comment (callaghanc) [reply spec/fixtures/model/force/lease/lease_domain.json]: yeah, for some reason all the already-existing lease domain fields were in camel case when i came to this story. i think you and @chrisdolendo did the previous lease work - does anyone recall why this is? i'm fine with changing them to snake case in this PR if there's no problem with doing that. even though we may ultimately decide that we want to convert domain to camel case down the line, until then it would be better to maintain consistency

File: spec/fixtures/model/force/lease/lease_salesforce_from_domain.json
Comment (callaghanc) [reply spec/fixtures/model/force/lease/lease_salesforce_from_domain.json:1]: we specify both the origin format and the result format, because the content of a translated object will depend on where that object originated from. objects can have fields that are only present when they are fetched from salesforce, or fields that are only present when they are fetched from the custom api. if i fetch, say, an application from the custom api, and then translate it to domain, it will only have the domain correlates of the custom api fields. if i fetch that same application via SOQL and translate it to domain, it will have the domain field correlates for those salesforce fields, which are a slightly different set of fields from what is returned by the custom api. this is a lot more relevant for application and preference objects, but i wanted to maintain consistency when testing all objects.

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (callaghanc) [reply spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: yeah i've also been thinking that - i wonder if there's any problem with doing that breaking up now? @fddayan @chrisdolendo how would you guys feel about breaking up the supp app e2e file now?

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (callaghanc) [reply spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: i wanted to keep them in the order that they're filled in on the page - especially for debugging, it makes it a lot easier to see "ok, we wanted to use this value for this field, then this value for the next....and the test got to this field, then the next..."

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (callaghanc) [reply spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: because it's not actually the last table row, it's only the last visible row - there's a hidden row right after it that is its corresponding expandable panel

File: spec/fixtures/model/force/lease/lease_domain.json
Comment (akegan) [reply spec/fixtures/model/force/lease/lease_domain.json]: yeah probably my fault . If you're up for it I think we should make it consistent. If you'd prefer, I can followup in a new PR after this one

**Review by akegan — APPROVED** [review]
looks good. 

Only outstanding things are possibly
1. Updating lease to have snake case domain model
1. Splitting up or organizing supplemental app tests

But it all looks good to me!

File: spec/fixtures/model/force/lease/lease_domain.json
Comment (callaghanc) [reply spec/fixtures/model/force/lease/lease_domain.json]: i'll do it in this PR, i got most of it done already during retro and sprint planning 😀

Comment (callaghanc) [reply spec/fixtures/model/force/lease/lease_domain.json]: done!

**Comment by callaghanc** [comment]
I've updated this PR with standardized lease domain field names. This PR is awaiting some responses from @fddayan and/or @chrisdolendo on some questions.

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (fddayan) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: I would keep a method for each action. I agree we would DRY up those methods.

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (fddayan) [reply spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: 👍I'm ok with splitting the file if it's growing. We could re think the naming convention we use for e2e tests. Instead of using the Page as the main test we can can `SupplementalApplicationPage/feature_name.e2e.test` 
like 
`SupplementalApplicationPage/manage_rental_asistances.e2e.test` .... or any name

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (callaghanc) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: @akegan Fed has added a commit showing a way to draw out the common parts of the rental assistance action functions - please review and let us know if you'd like to use that new way or keep it as it was before.

File: spec/javascript/e2e/SupplementalApplicationPage.e2e.js
Comment (callaghanc) [reply spec/javascript/e2e/SupplementalApplicationPage.e2e.js]: @akegan Fed and I have broken apart the supp app e2e tests in a couple of recent commits - please review.

**Comment by callaghanc** [comment]
@akegan Fed and I have added a few commits to this PR. In addition to addressing the two items you listed as outstanding in your last comment, these commits also do a bit of lease mapping work and correct a field mapping issue with preferences. So please review the commits since you last reviewed and let us know your thoughts.

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (akegan) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: I'm ambivalent! Either way works. Up to you guys!

File: spec/javascript/e2e/SupplementalApplicationPage/preferences.e2e.js
Comment (akegan) [inline spec/javascript/e2e/SupplementalApplicationPage/preferences.e2e.js:1]: small nit: filenames aren't super consistent for these new broken-out things.

action_buttons
preferences
manage_rental_assistance
status_history

I think it might be a bit clearer if we pick a convention and stick with it. 

I'd prefer if we keep them to domain-related topics, e.g. 
-preferences
-rental_assistance
-status_update (combine status history and action_buttons)

File: app/javascript/components/supplemental_application/sections/RentalAssistance.js
Comment (callaghanc) [inline app/javascript/components/supplemental_application/sections/RentalAssistance.js:32]: @jrubenoff I've changed the capitalization on these two options back to their original Salesforce picklist capitalizations, so that they will exactly match the picklist values.

File: app/javascript/components/supplemental_application/SupplementalApplicationPage.js
Comment (callaghanc) [reply app/javascript/components/supplemental_application/SupplementalApplicationPage.js:263]: Ok let's leave it as it is now

File: spec/javascript/e2e/SupplementalApplicationPage/preferences.e2e.js
Comment (callaghanc) [reply spec/javascript/e2e/SupplementalApplicationPage/preferences.e2e.js:1]: sounds good to me, done!

File: app/javascript/components/supplemental_application/sections/RentalAssistance.js
Comment (jrubenoff) [reply app/javascript/components/supplemental_application/sections/RentalAssistance.js:32]: That's unfortunate, but OK!

**Comment by akegan** [comment]
@callaghanc I've merged latest master into this 🤞the tests will pass and we can merge this in

**Pull request merged**
